app/models/user.rb

```
@@ -381,6 +383,56 @@ def only_card_grant_user?
     card_grants.size >= 1 && events.size == 0
   end
 
+  def backup_codes_enabled?
+    active_backup_codes.any?
+  end
+
+  def generate_backup_codes!
+    backup_codes.previewed.destroy_all
+
+    codes = []
+    while codes.size < 10
+      code = SecureRandom.alphanumeric(10)
+      next if codes.include?(code)
+
+      begin
+        backup_codes.create!(code: code)
+      rescue ActiveRecord::RecordInvalid
+        # if the code is already in use, skip it
+        next
+      end
```

**concern**: Given the salt + hash are we likely to ever hit the `rescue` branch?

---

If something does go wrong we probably don't want to tank the whole generation process, so it might still be good to have that rescue branch in there?

Not very sure about this tbh

---

> If something does go wrong we probably don't want to tank the whole generation process

If something does happen it would be highly unexpected and TBH I'd rather fail loudly than try to recover.

---

Probably not, I'll remove the rescue statement